README.md
# COVID-19 Cases Worldwide Dataset from ECDC

## What is this?

The goal of this project is to provide downloader and loader for geographic distribution of COVID-19 cases worldwide dataset from [ECDC's website](https://www.ecdc.europa.eu/en/publications-data/download-todays-data-geographic-distribution-covid-19-cases-worldwide).

*We highly recommend to use this or similar script or module instead of downloading the dataset each time from the server of ECDC to spare data usage and keep bandwidth for other researchers and developers as well.* Both solutions here checks whether today's data is downloaded yet or not and downloads only in case if it is needed.

## Parts of the project

### Python script to use as downloader or as a module

The file ` ecdcloader.py ` is a python script designed to be used as a module or as a downloader at the same time.

#### As module

Simpla paste the file ` ecdcloader.py ` into your working directory and add ` import ecdcloader ` line to your code. After importing the module, the variable ` dataset ` will contain the actual data in the form of ` pandas.DataFrame() `.

#### As script

Simply run ` ecdcloader.py ` and you'll get the newest file.

#### Fine tuning

If you want to use different settings

- ` LOAD_LASTDATA_IF_DOWNLOAD_FAILS ` Constant to specify fallback behavior. If True, tries to get data from yesterday. (Default value: *True*)
- ` TRYING_LIMIT ` Number of trying if error occurs during download. (Default value: *3*)
- ` LINK_CORE ` Root link of ECDC website's download area. (Default value: *https://www.ecdc.europa.eu/sites/default/files/documents/*)
- ` FILENAME_CORE ` Set xlsx file prefix (Default value: *COVID-19-geographic-disbtribution-worldwide-*)

### Python script and its binary to use as executable

The windows executable ` getdailyecdc.exe ` is built from ` getdailyecdc.py ` you can check it's virus-free status at [jotty's page](https://virusscan.jotti.org/en-US/filescanjob/1xzi4o68p0) or you can check the result of the query [here](https://github.com/hyperrixel/COVID19CasesWorldwideDatasetFromECDC/blob/master/virus_free_proof.jpg).

The executable accepts command-line parameters.

#### Command-line parameters

- ` -h ` ` --help ` Displays help screen.
- ` -f DAYS ` ` --fallback-days DAYS ` Sets count of fallback days. Default is ` 1 `. To pass fallback set it to ` 0 `.
- ` -p NAME ` ` --prefix NAME ` Sets xlsx file prefix. Default is "COVID-19-geographic-disbtribution-worldwide-".
- ` -r NAME ` ` --root-link NAME ` Sets download link root. Default is "https://www.ecdc.europa.eu/sites/default/files/documents/".
- ` -t TRIES ` ` --try-limit TRIES ` Sets count of download try-limits Default is ` 3 `. Use higher value in case of very bad connection circumstances.
- ` -v ` ` --version ` Shows program's version number and exits

![Screenshot - ommand line parameters in real life](ecdc-data.png)

## Future plans

Support of other filetypes: ` csv ` ` xml ` ` json `


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [1.0.0] - 2020-03-27
### Added
- Download the daily ' .xlsx ' file from [ECDC](https://www.ecdc.europa.eu/en)'s
server
- Load latest dataset into a variable on the end of process in the python version
- Use he software in script-mode and module-mode also
- Option to load the yesterday's data if error occurs during downloading the
actual dataset. If the script does not find the stored version of yesterday's
data, it download the missing dataset from ECDC
- Error handling when there is no data to download or to process
- File extension is hardcoded as ' .xlsx ', that means there is no additional
support for other filetypes
- Add trying limit into downloading part to balance between maximize the ability
to download the file and minimize the unneccesary loading of remote server
- Use printsummary function to give back a human-readable printed feedback
about working flow of software
- Create and support executable version
- Executable version support starting with argparser
- Start following [SemVer](https://semver.org/spec/v2.0.0.html)
- Start using CHANGELOG as CHANGELOG.md
- Start using README as README.md
- Start using REQUIREMENTS as requirements.txt


ecdcloader.py
"""
        _                 _
       (_)               | |
 _ __   _  __  __   ___  | |
| '__| | | \ \/ /  / _ \ | |
| |    | |  >  <  |  __/ | |
|_|    |_| /_/\_\  \___| |_|

COVID-19-data-from-ECDC
-----------------------

The goal of this project is to provide python based downloader and loader for
geographic distribution of COVID-19 cases worldwide dataset from ECDC's website.

Copyright rixel 2020
Distributed under the Boost Software License, Version 1.0.
See accompanying file LICENSE or a copy at https://www.boost.org/LICENSE_1_0.txt
"""



# Special constants for modul style use.
__author__ = 'rixel'
__copyright__ = "Copyright 2020, COVID-19-data-from-ECDC Project"
__credits__ = ['rixel']
__license__ = 'BOOST 1.0'
__version__ = '1.0'
__status__ = 'Beta'



# Import section
# Imports from standard library
from datetime import date, timedelta
import requests
from os.path import isfile

# Imports of external libraries
import pandas as pd



# Constant to specify fallback behavior.
# If True, tries to get data from yesterday.
LOAD_LASTDATA_IF_DOWNLOAD_FAILS = True

# Constant to set number of download tries.
TRYING_LIMIT = 3

# Constant to root link of ECDC website's download area.
LINK_CORE = 'https://www.ecdc.europa.eu/sites/default/files/documents/'

# Constant to set xlsx file prefix
FILENAME_CORE = 'COVID-19-geographic-disbtribution-worldwide-'



def generatelinkandname(day_minus=None):
    """
    Generates link and filename strings to given day
    ------------------------------------------------
    @Params: (int)      [optional] Count of day before today. If omitted today's
                        date will be used.
    @Return: (tuple)    Tuple of download-link and filename strings.
    """

    global LINK_CORE
    global FILENAME_CORE

    if day_minus == None:
        day = date.today().strftime("%Y-%m-%d")
    else:
        day = (date.today() - timedelta(days = day_minus)).strftime("%Y-%m-%d")
    link = '{}{}{}.xlsx'.format(LINK_CORE, FILENAME_CORE, day)
    name = './{}{}.xlsx'.format(FILENAME_CORE, day)
    return link, name



def getfile(link):
    """
    Gets file from link and prints the status of the process
    --------------------------------------------------------
    @Params: link (string)      Link to get the file from.
    @Return: (data|Nonetype)    Data in case of success else None.
    """

    global TRYING_LIMIT

    file_data = None
    trying_count = 1
    print('')
    while file_data == None and trying_count <= TRYING_LIMIT:
        try:
            print('\rTrying to get {}... ({}/{})'
                  .format(link, trying_count, TRYING_LIMIT), end='')
            back_data = requests.get(link)
            if back_data.status_code == 200:
                file_data = back_data.content
        except:
            pass
        trying_count += 1
        if file_data == None:
            print(' FAILED!')
        else:
            print(' SUCCESS.')
    return file_data



def printsummary(text):
    """
    Prints summary and text
    -----------------------
    @Params: text   (string)    Text to print after summary header.
    """

    print('Summary:')
    print('--------')
    print(text)



def simpleget(minus_day=None):
    """
    Downloads data and generates link and filename strings to given day
    -------------------------------------------------------------------
    @Params: (int)      [optional] Count of day before today. If omitted today's
                        date will be used.
    @Return: (tuple)    Tuple of downloaded data, download-link and filename.
    """

    link, name = generatelinkandname(minus_day)
    data = getfile(link)
    return data, link, name



link, name = generatelinkandname()
if not isfile(name):
    data = getfile(link)
    if data != None:
        print('Saving data to {}...'.format(name), end='')
        with open(name, 'wb') as outstream:
            outstream.write(data)
        print(' DONE.')
        dataset = pd.read_excel(name)
        printsummary('Download finished successfully.\nYou can use the new file:\n{}'
                     .format(name))
    elif LOAD_LASTDATA_IF_DOWNLOAD_FAILS:
        link, name = generatelinkandname(1)
        if isfile(name):
            dataset = pd.read_excel(name)
            printsummary('Failed to download daily file.\nYou can use file and data from yesterday:\n{}'
                         .format(name))
        else:
            data = getfile(link)
            if data != None:
                print('Saving data to {}...'.format(name), end='')
                with open(name, 'wb') as outstream:
                    outstream.write(data)
                print(' DONE.')
                dataset = pd.read_excel(name)
                printsummary('Failed to download daily file.\nNo dataset from yesterday present but it is downloaded successfully.\n'
                             + 'You can use file and data from yesterday:\n{}'
                             .format(name))
            else:
                printsummary('Failed to download daily file.\nFallback to yesterday wasn\'t successful too.\n'
                             + 'To solve this issue try simpleget(minus_day) function with more than 1 minus day parameter.\n')
    else:
        printsummary('Failed to download daily file.\nNo fallback is set, nothing downloaded.\n' +
                     'Set LOAD_LASTDATA_IF_DOWNLOAD_FAILS to True to download data from yesterday.')
else:
    dataset = pd.read_excel(name)
    printsummary('Today\'s data already downloaded.\nYou can use the file:\n{}'
                 .format(name))


getdailyecdc.py
"""
        _                 _
       (_)               | |
 _ __   _  __  __   ___  | |
| '__| | | \ \/ /  / _ \ | |
| |    | |  >  <  |  __/ | |
|_|    |_| /_/\_\  \___| |_|

COVID-19-data-from-ECDC
-----------------------

The goal of this project is to provide python based downloader and loader for
geographic distribution of COVID-19 cases worldwide dataset from ECDC's website.

Copyright rixel 2020
Distributed under the Boost Software License, Version 1.0.
See accompanying file LICENSE or a copy at https://www.boost.org/LICENSE_1_0.txt
"""



# Import section
# Imports from standard library
from argparse import ArgumentParser, SUPPRESS
from datetime import date, timedelta
import requests
from os.path import isfile



# Constant to set version [Major, Minor, Build, Release type]
VERSION = [1, 0, 0, 'beta']

# Constant to specify fallback behavior.
# If >0, tries to get data from count of previous days, 0 is for no fallback.
LOAD_LASTDATA_IF_DOWNLOAD_FAILS = 1

# Constant to set number of download tries.
TRYING_LIMIT = 3

# Constant to root link of ECDC website's download area.
LINK_CORE = 'https://www.ecdc.europa.eu/sites/default/files/documents/'

# Constant to set xlsx file prefix
FILENAME_CORE = 'COVID-19-geographic-disbtribution-worldwide-'



def generatelinkandname(day_minus=None):
    """
    Generates link and filename strings to given day
    ------------------------------------------------
    @Params: (int)      [optional] Count of day before today. If omitted today's
                        date will be used.
    @Return: (tuple)    Tuple of download-link and filename strings.
    """

    global args

    if day_minus == None:
        day = date.today().strftime("%Y-%m-%d")
    else:
        day = (date.today() - timedelta(days = day_minus)).strftime("%Y-%m-%d")
    link = '{}{}{}.xlsx'.format(args.root_link, args.prefix, day)
    name = './{}{}.xlsx'.format(args.prefix, day)
    return link, name



def getfile(link):
    """
    Gets file from link and prints the status of the process
    --------------------------------------------------------
    @Params: link (string)      Link to get the file from.
    @Return: (data|Nonetype)    Data in case of success else None.
    """

    global args

    file_data = None
    trying_count = 1
    print('')
    while file_data == None and trying_count <= args.try_limit:
        try:
            print('\rTrying to get {}... ({}/{})'
                  .format(link, trying_count, args.try_limit), end='')
            back_data = requests.get(link)
            if back_data.status_code == 200:
                file_data = back_data.content
        except:
            pass
        trying_count += 1
        if file_data == None:
            print(' FAILED!')
        else:
            print(' SUCCESS.')
    return file_data



def getversion():
    """
    Returns version variable as a string
    ------------------------------------
    @Return: (string)   Version string in format <Major>.<Minor>.<Build>.<Release-type>.
    """

    global VERSION

    if len(VERSION) == 3:
        return '{}.{}.{}'.format(VERSION[0], VERSION[1], VERSION[2])
    else:
        return '{}.{}.{}-{}'.format(VERSION[0], VERSION[1], VERSION[2], VERSION[3])



def printsummary(text):
    """
    Prints summary and text
    -----------------------
    @Params: text   (string)    Text to print after summary header.
    """

    print('Summary:')
    print('--------')
    print(text)



parser = ArgumentParser(description='Downloader for geographic distribution of COVID-19 cases worldwide dataset from ECDC\'s website', epilog='')
parser.add_argument('-f', '--fallback-days', default=LOAD_LASTDATA_IF_DOWNLOAD_FAILS,
                    help='set count of fallback days (default is {}, 0 is no fallback)'
                    .format(LOAD_LASTDATA_IF_DOWNLOAD_FAILS), metavar='DAYS', type=int)
parser.add_argument('-p', '--prefix', default=FILENAME_CORE,
                    help='set xlsx file prefix (default is "{}")'.format(FILENAME_CORE),
                    metavar='NAME')
parser.add_argument('-r', '--root-link', default=LINK_CORE,
                    help='set download link root (default is "{}")'.format(LINK_CORE),
                    metavar='NAME')
parser.add_argument('-t', '--try-limit', default=TRYING_LIMIT,
                    help='set count of download try-limits (default is 3)', metavar='TRIES', type=int)
parser.add_argument('-v', '--version', action='version', default=SUPPRESS, version=getversion())
args = parser.parse_args()

link, name = generatelinkandname()
if not isfile(name):
    data = getfile(link)
    if data != None:
        print('Saving data to {}...'.format(name), end='')
        with open(name, 'wb') as outstream:
            outstream.write(data)
        print(' DONE.')
        printsummary('Download finished successfully.\nYou can use the new file:\n{}'
                     .format(name))
    elif args.fallback_days > 0:
        no_result = True
        for i in range(1, args.fallback_days + 1):
            link, name = generatelinkandname(i)
            if isfile(name):
                printsummary('Failed to download daily file.\nYou can use nearest available file and data from here:\n{}'
                             .format(name))
                no_result = False
                break
            else:
                data = getfile(link)
                if data != None:
                    print('Saving data to {}...'.format(name), end='')
                    with open(name, 'wb') as outstream:
                        outstream.write(data)
                    print(' DONE.')
                    printsummary('Failed to download daily file.\nNo datasets before this day present but it is downloaded successfully.\n'
                                 + 'You can use file and data from here:\n{}'
                                 .format(name))
                    no_result = False
                    break
        if no_result:
            printsummary('Failed to download daily file.\nFallback to previous days also failed in given interval.\n'
                         + 'To solve this issue try to set higher fallback_days value.\n')
    else:
        printsummary('Failed to download daily file.\nNo fallback is set, nothing downloaded.\n' +
                     'Set fallback_days different than 0 to download data from previous day(s).')
else:
    printsummary('Today\'s data already downloaded.\nYou can use the file:\n{}'
                 .format(name))
